language_detector/main.py

```diff
@@ -1,9 +1,61 @@
+# -*- coding: utf-8 -*-
 """This is the entry point of the program."""
 
 from .languages import LANGUAGES
+import string
+import collections
+import re
 
+# languages = [{'name':"spanish",'common_words':'['el',...]}, {'name':"english",...}]
 
 def detect_language(text, languages):
     """Returns the detected language of given text."""
-    # implement your solution here
-    pass
+    
+    # remove any punctuation and convert input text to lowercase
+    bad_chars = '[' + string.punctuation + ']'
+    clean_text = re.sub(bad_chars, '', text)
+    clean_text = clean_text.lower()
+    text_words = clean_text.split()
+    
+    # we will use sets to find the shared unique words between the input text and language lists
+    word_set = set(text_words)
+    
+    # use a dict with key = # matching words and value = set of language names
+    # this is used to return the highest scoring language(s) at the end
+    score_dict = collections.defaultdict(set)
+    
+    for language in languages:
+        language_words = language['common_words']
+        # both the input text and language list will be lowercase for comparison
+        language_words = [word.lower() for word in language_words]
+        language_set = set(language_words)
+        
+        # take the set intersection to get the common words between input and language list
+        common_set = word_set & language_set
+        
+        score_dict[len(common_set)].add(language['name'])
+    
+    # get the set of languages that had the top score
+    top_languages = score_dict[max(score_dict)]
+    
+    # if there is a single language with the best score, return that name as a string
+    if len(top_languages) == 1:
+        return next(iter(top_languages))
+    
+    # in case of a tie with multiple languages, return a set of language names    
+    return top_languages
```

Comment:
This looks like great code, I'm not sure all the cleaning was necessary, but otherwise it is simple